File: parser_.py
"""
GRAMMAR RULES:
        EXPR --> TERM | TERM + EXPR | TERM - EXPR
        TERM --> FACTOR | FACTOR * TERM | FACTOR / TERM
      FACTOR --> FACTOR_PRIME | FACTOR ^ FACTOR_PRIME
FACTOR_PRIME --> <ID, X> | CONSTANT | NUMBER | MATH(EXPR)
"""

from tokens import TokenType
from nodes import *


class Parser:
    def __init__(self, tokens):
        self.current_token = None
        self.tokens = iter(tokens)
        self.advance()

    def advance(self):
        try:
            self.current_token = next(self.tokens)
        except StopIteration:
            self.current_token = None

    def raise_error(self):
        raise Exception('Invalid Syntax!')

    def parse(self):
        if not self.current_token:
            return None
        result = self.expr()
        if self.current_token:
            self.raise_error()
        return result

    def expr(self):
        result = self.term()

        while self.current_token and self.current_token.type in (TokenType.PLUS, TokenType.MINUS):
            if self.current_token.type == TokenType.PLUS:
                self.advance()
                result = AddNode(result, self.term())
            elif self.current_token.type == TokenType.MINUS:
                self.advance()
                result = SubtractNode(result, self.term())
        return result

    def term(self):
        result = self.factor()
        while self.current_token and self.current_token.type in (TokenType.MULTIPLY, TokenType.DIVIDE):
            if self.current_token.type == TokenType.MULTIPLY:
                self.advance()
                result = MultiplyNode(result, self.factor())
            elif self.current_token.type == TokenType.DIVIDE:
                self.advance()
                result = DivideNode(result, self.factor())
        return result

    def factor(self):
        result = self.factor_prime()
        while self.current_token and self.current_token.type == TokenType.POWER:
            self.advance()
            result = PowerNode(result, self.factor_prime())
        return result

    def factor_prime(self):
        token = self.current_token
        if token.type == TokenType.NUMBER:
            self.advance()
            return NumberNode(token.value)
        if token.type == TokenType.VAR:
            self.advance()
            return VarNode()
        if token.type == TokenType.CONSTANT:
            self.advance()
            return ConstantNode(token.value)
        if token.type == TokenType.MATH_FUNCTION:
            math_token = token
            self.advance()
            self.advance()  # skip LPAREN
            result = self.expr()
            self.advance()  # skip RPAREN
            return MathNode(math_token.value, result)
        if token.type == TokenType.PLUS:
            self.advance()
            return PlusNode(self.factor_prime())
        if token.type == TokenType.MINUS:
            self.advance()
            return PlusNode(self.factor_prime())
        self.raise_error()

File: Java Version/math_interpreter/src/com/company/Parser.java
package com.company;

/*
      EXPR         --> TERM | TERM + EXPR | TERM - EXPR
      TERM         --> FACTOR | FACTOR * TERM | FACTOR / TERM
      FACTOR       --> FACTOR_PRIME | FACTOR ^ FACTOR_PRIME
      FACTOR_PRIME --> VAR: X | CONSTANT | NUMBER | MATH(EXPR) | + FACTOR_PRIME | - FACTOR_PRIME | [LPAREN] EXPR [RPAREN]
 */

import com.company.nodes.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Parser {
    private final ArrayList<Token> tokens;
    private int index;
    private Token currentToken;

    Parser(ArrayList<Token> tokens)
    {
        this.tokens = tokens;
        this.index = -1;
        this.currentToken = null;
        this.advance();
    }

    public void raiseError() throws Exception {
        throw new Exception(String.format("Invalid syntax: %s", this.currentToken));
    }

    private void advance()
    {
        if(++this.index >= tokens.size())
            this.currentToken = null;
        else this.currentToken = tokens.get(this.index);
    }

    private Node expr() throws Exception {
        Node result = this.term();
        Set<TokenType> acceptedTokens = Set.of(TokenType.PLUS, TokenType.MINUS);
        while (this.currentToken != null && acceptedTokens.contains(this.currentToken.tokenType)) {
            if(this.currentToken.tokenType == TokenType.PLUS)
            {
                this.advance();
                result = new AddNode(result, this.term());
            }
            else if(this.currentToken.tokenType == TokenType.MINUS)
            {
                this.advance();
                result = new SubtractNode(result, this.term());
            }
        }
        return result;
    }

    private Node factor() throws Exception {
        Node result = this.factorPrime();
        while(this.currentToken != null && this.currentToken.tokenType == TokenType.POWER)
        {
            this.advance();
            result = new PowerNode(result, this.factor());
        }
        return result;
    }

    private Node factorPrime() throws Exception {
        Token token = this.currentToken;
        Node result = null;
        if(token.tokenType == TokenType.LPAREN)
        {
            this.advance();
            result = this.expr();
            if(token.tokenType != TokenType.RPAREN)
                raiseError();
            this.advance();
        }
        else if(token.tokenType == TokenType.VAR)
        {
            this.advance();
            result = new VarNode();
        }
        else if(token.tokenType == TokenType.NUMBER)
        {
            this.advance();
            result = new NumberNode(Double.parseDouble(token.value));
        }
        else if(token.tokenType == TokenType.CONSTANT)
        {
            this.advance();
            result = new ConstantNode(token.value);
        }
        else if(token.tokenType == TokenType.MATH_FUNCTION)
        {
            Token mathToken = this.currentToken;
            this.advance();
            this.advance(); // LPAREN
            result = this.expr();
            this.advance(); // RPAREN
            result = new MathNode(mathToken, result);
        }
        else if(token.tokenType == TokenType.PLUS)
        {
            this.advance();
            result = new PlusNode(this.factorPrime());
        }
        else if(token.tokenType == TokenType.MINUS)
        {
            this.advance();
            result = new MinusNode(this.factorPrime());
        }
        return result;
    }


    private Node term() throws Exception {
        Node result = this.factor();
        Set<TokenType> acceptedTokens = Set.of(TokenType.MULTIPLY, TokenType.DIVIDE);
        while(this.currentToken != null && acceptedTokens.contains(this.currentToken.tokenType))
        {
            if(this.currentToken.tokenType == TokenType.MULTIPLY)
            {
                this.advance();
                result = new MultiplyNode(result, this.term());
            }
            else if(this.currentToken.tokenType == TokenType.DIVIDE)
            {
                this.advance();
                result = new DivideNode(result, this.term());
            }
        }
        return result;
    }

    public Node parse() throws Exception
    {
        if(this.currentToken == null)
            return null;
        Node result = this.expr();
        System.out.println(this.currentToken);
        if(this.currentToken != null)
            raiseError();
        return result;
    }

}
